spec/jobs/reports/drop_off_report_spec.rb

```diff
@@ -7,11 +7,77 @@
     JSON.parse '[{"emails":["ursula@example.com"],
        "issuers":["urn:gov:gsa:openidconnect.profiles:sp:sso:agency_name:app_name"]}]'
   end
+  let(:empty_emailable_report) do
+    report = Reporting::DropOffReport.new(
+      issuers: 'urn:gov:gsa:openidconnect.profiles:sp:sso:agency_name:app_name',
+      time_range: report_date.all_week,
+      slice: 1.week,
+    )
+
+    report.as_emailable_reports
+  end
 
   before do
     allow(IdentityConfig.store).to receive(:drop_off_report_config).and_return(report_config)
   end
 
+  describe 'with empty logs' do
+    before do
+      stub_cloudwatch_logs([])
+    end
+
+    it 'sends a ReportMailer with data that matches an empty Reporting::DropOffReport' do
+      expect(ReportMailer).to receive(:tables_report).once.with(
+        email: 'ursula@example.com',
+        subject: "Drop Off Report - #{report_date.to_date}",
+        reports: satisfy do |value|
+          expect(value.first).to eq(empty_emailable_report.first)
+          expect(value.second).to eq(empty_emailable_report.second)
+          expect(value.last.to_json).to eq(empty_emailable_report.last.to_json)
+          expect(value.count).to be(3)
+        end,
+        message: "<h2>\n  Drop Off Report\n</h2>\n",
+        attachment_format: :csv,
+      ).and_call_original
+
+      subject.perform(report_date)
+    end
+
+    it 'accepts a string for issuers instead of an array' do
+      report_config[0]['issuers'] = 'urn:gov:gsa:openidconnect.profiles:sp:sso:agency_name:app_name'
+      subject.perform(report_date)
+    end
+  end
+
+  describe 'with some log data' do
+    before do
+      stub_cloudwatch_logs(
+        [
+          # gets through phone finder, then drops
+          { 'user_id' => 'user2', 'name' => 'IdV: doc auth welcome visited' },
+          { 'user_id' => 'user2', 'name' => 'IdV: doc auth welcome submitted' },
+          { 'user_id' => 'user2', 'name' => 'IdV: doc auth image upload vendor submitted' },
+          { 'user_id' => 'user2', 'name' => 'IdV: doc auth document_capture visited' },
+          { 'user_id' => 'user2', 'name' => 'IdV: doc auth ssn visited' },
+          { 'user_id' => 'user2', 'name' => 'IdV: doc auth verify visited' },
+          { 'user_id' => 'user2', 'name' => 'IdV: doc auth verify submitted' },
+          { 'user_id' => 'user2', 'name' => 'IdV: phone of record visited' },
+        ],
+      )
+    end
+
+    it 'sends a mailer with the log data' do
+      subject.perform(report_date)
+      sent_mail = ActionMailer::Base.deliveries.last
+      csv_data = CSV.parse(sent_mail.parts.last.body.to_s)
+      phone_finder_row = csv_data.find_index { |row| row[0] == 'Phone finder (page view)' }
+
+      expect(csv_data[0][2]).to eq('Users lost') # make sure the headers are still in the same order
+      expect(csv_data[phone_finder_row][2]).to eq('0')
+      expect(csv_data[phone_finder_row + 1][2]).to eq('1')
```

Comment:
we can have the `CSV` lib parse headers and use those to make at lot of this code more readable. I would not necessarily assert on the order of headers, but that's what the old code did so I translated it here

```suggestion
      csv_data = CSV.parse(sent_mail.parts.last.body.to_s, headers: true)
      expect(csv_data.headers).to eq(
        [
          'Step',
          'Unique user count',
          'Users lost',
          'Dropoff from last step',
          'Users left from start'
        ],
      )

      phone_finder_row = csv_data.find { |row| row['Step'] == 'Phone finder (page view)' }
      expect(phone_finder_row['Users lost']).to eq('0')
      expect(phone_finder_row['Dropoff from last step']).to eq('1')
```